### PR TITLE
OVAL in foreman_openscap

### PR DESCRIPTION
**Pull request by xprazak2**
Contains #456 and #457 on top of each other for easier testing.

It should be possible to create oval policy with oval content and assign it to host(group)s.

OVAL content from RH stream is bzipped, I managed to upload with:

```
echo "{ \"oval_content\": { \"scap_file\": \"$(base64 -w 0 ~/Downloads/whatever.oval.xml.bz2)\"  } }" | curl -H "Accept: application/json" -H "Content-Type: application/json" -k -X PUT -u admin:changeme https://centos-devel.example.com/api/v2/compliance/oval_contents/3 -d @-  | json_reformat
``` 

though we might need to think about whether we want to unzip before uploading.

### CODE REVIEW

**Comment by xprazak2** [comment]
Foreman is now able to attach CVEs to hosts and set OVAL status.

**Comment by xprazak2** [comment]
I added changes to keep track of CVEs without Errata, I also updated the report template on foreman side.

**Comment by xprazak2** [comment]
[test foreman_openscap]

File: app/models/foreman_openscap/oval_status.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
+module ForemanOpenscap
+  class OvalStatus < ::HostStatus::Status
+    PATCHED = 0
+    VULNERABLE = 1
```
Comment (ares) [inline app/models/foreman_openscap/oval_status.rb:4]: Ideally we'd have 3 different states

OK - no CVEs found (0 CVEs records for the host)
Warning - CVEs found, no errata available for any of them
Error - CVEs found, errata for them exist (patch available, but not applied)

**Review by ares — CHANGES_REQUESTED** [review]
I have tested the functionality end to end and it works pretty well! Leaving a few smaller comments here.

File: app/models/foreman_openscap/oval_status.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+    def relevant?(options = {})
+      host.combined_oval_policies.any?
+    end
+
+    def to_status(options = {})
+      return PATCH_AVAILABLE if host.cves_with_errata.any?
+      return VULNERABLE if host.cves_without_errata.any?
+      PATCHED
+    end
```
Comment (xprazak2) [inline app/models/foreman_openscap/oval_status.rb:43]: If there is at least one CVE with existing errata, status will be error.

**Comment by ares** [comment]
[test foreman_openscap], code looks good now, I'm retesting the functionality, failures seemed unrelated

File: app/controllers/api/v2/compliance/oval_reports_controller.rb [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        param :cname, :identifier, :required => true
+        param :oval_policy_id, :identifier, :required => true
+        param :date, :identifier, :required => true
+
+        def create
+          ForemanOpenscap::Oval::Cves.new.create(@host, params.to_unsafe_h)
+          if @host.errors.any?
+            upload_fail @host.errors.full_messages
+          else
+            @host.refresh_statuses([HostStatus.find_status_by_humanized_name("oval")])
```
Comment (ares) [inline app/controllers/api/v2/compliance/oval_reports_controller.rb:21]: `@host.refresh_statuses([HostStatus.find_status_by_humanized_name("oval")])` this no longer works, it should be
`@host.refresh_statuses([ForemanOpenscap::OvalStatus])`

**Review by ares — CHANGES_REQUESTED** [review]
I found that parsing the report creates SQL per each CVE. We don't have to address it here, but some performance tweaking will be necessary. Here's the sample of the SQL log.

```
2021-04-28T10:53:26 [D|sql|e4b3d4a0]   ↳ /home/ares/Projekty/Zdrojaky/foreman_openscap/app/services/foreman_openscap/oval/cves.rb:10:in `block (2 levels) in create'
2021-04-28T10:53:26 [D|sql|e4b3d4a0]   ForemanOpenscap::Cve Load (0.9ms)  SELECT "foreman_openscap_cves".* FROM "foreman_openscap_cves" WHERE "foreman_openscap_cves"."ref_id" = $1 AND "foreman_openscap_cves"."ref_url" = $2 AND "foreman_openscap_cves"."has_errata" = $3 AND "foreman_openscap_cves"."definition_id" = $4 LIMIT $5  [["ref_id", "CVE-2014-3537"], ["ref_url", "https://access.redhat.com/security/cve/CVE-2014-3537"], ["has_errata", true], ["definition_id", "oval:com.redhat.rhba:def:20150386"], ["LIMIT", 1]]
2021-04-28T10:53:26 [D|sql|e4b3d4a0]   ↳ /home/ares/Projekty/Zdrojaky/foreman_openscap/app/services/foreman_openscap/oval/cves.rb:10:in `block (2 levels) in create'
2021-04-28T10:53:26 [D|sql|e4b3d4a0]   ForemanOpenscap::Cve Load (1.0ms)  SELECT "foreman_openscap_cves".* FROM "foreman_openscap_cves" WHERE "foreman_openscap_cves"."ref_id" = $1 AND "foreman_openscap_cves"."ref_url" = $2 AND "foreman_openscap_cves"."has_errata" = $3 AND "foreman_openscap_cves"."definition_id" = $4 LIMIT $5  [["ref_id", "CVE-2014-5029"], ["ref_url", "https://access.redhat.com/security/cve/CVE-2014-5029"], ["has_errata", true], ["definition_id", "oval:com.redhat.rhba:def:20150386"], ["LIMIT", 1]]
2021-04-28T10:53:26 [D|sql|e4b3d4a0]   ↳ /home/ares/Projekty/Zdrojaky/foreman_openscap/app/services/foreman_openscap/oval/cves.rb:10:in `block (2 levels) in create'
2021-04-28T10:53:26 [D|sql|e4b3d4a0]   ForemanOpenscap::Cve Load (0.9ms)  SELECT "foreman_openscap_cves".* FROM "foreman_openscap_cves" WHERE "foreman_openscap_cves"."ref_id" = $1 AND "foreman_openscap_cves"."ref_url" = $2 AND "foreman_openscap_cves"."has_errata" = $3 AND "foreman_openscap_cves"."definition_id" = $4 LIMIT $5  [["ref_id", "CVE-2014-5030"], ["ref_url", "https://access.redhat.com/security/cve/CVE-2014-5030"], ["has_errata", true], ["definition_id", "oval:com.redhat.rhba:def:20150386"], ["LIMIT", 1]]
2021-04-28T10:53:26 [D|sql|e4b3d4a0]   ↳ /home/ares/Projekty/Zdrojaky/foreman_openscap/app/services/foreman_openscap/oval/cves.rb:10:in `block (2 levels) in create'
2021-04-28T10:53:26 [D|sql|e4b3d4a0]   ForemanOpenscap::Cve Load (0.9ms)  SELECT "foreman_openscap_cves".* FROM "foreman_openscap_cves" WHERE "foreman_openscap_cves"."ref_id" = $1 AND "foreman_openscap_cves"."ref_url" = $2 AND "foreman_openscap_cves"."has_errata" = $3 AND "foreman_openscap_cves"."definition_id" = $4 LIMIT $5  [["ref_id", "CVE-2014-5031"], ["ref_url", "https://access.redhat.com/security/cve/CVE-2014-5031"], ["has_errata", true], ["definition_id", "oval:com.redhat.rhba:def:20150386"], ["LIMIT", 1]]
2021-04-28T10:53:26 [D|sql|e4b3d4a0]   ↳ /home/ares/Projekty/Zdrojaky/foreman_openscap/app/services/foreman_openscap/oval/cves.rb:10:in `block (2 levels) in create'
2021-04-28T10:53:26 [D|sql|e4b3d4a0]   ForemanOpenscap::Cve Load (0.9ms)  SELECT "foreman_openscap_cves".* FROM "foreman_openscap_cves" WHERE "foreman_openscap_cves"."ref_id" = $1 AND "foreman_openscap_cves"."ref_url" = $2 AND "foreman_openscap_cves"."has_errata" = $3 AND "foreman_openscap_cves"."definition_id" = $4 LIMIT $5  [["ref_id", "CVE-2014-1545"], ["ref_url", "https://access.redhat.com/security/cve/CVE-2014-1545"], ["has_errata", true], ["definition_id", "oval:com.redhat.rhba:def:20150364"], ["LIMIT", 1]]
2021-04-28T10:53:26 [D|sql|e4b3d4a0]   ↳ /home/ares/Projekty/Zdrojaky/foreman_openscap/app/services/foreman_openscap/oval/cves.rb:10:in `block (2 levels) in create'
2021-04-28T10:53:26 [D|sql|e4b3d4a0]    (3.8ms)  SELECT "foreman_openscap_cves"."id" FROM "foreman_openscap_cves" INNER JOIN "foreman_openscap_host_cves" ON "foreman_openscap_cves"."id" = "foreman_openscap_host_cves"."cve_id" WHERE "foreman_openscap_host_cves"."host_id" = $1  [["host_id", 76]]
```

Also it creates instance of `HostStatus::Status` instead of `ForemanOpenscap::OvalStatus` which then fails to display. That's the last thing that needs fixing (added a comment in the code)

**Comment by xprazak2** [comment]
I updated the status refresh, I'll tweak the performance separately. Some tests from core will not pass due to interaction with rex, that is why I opened the PR to move to gh actions for testing (which can be done later).

**Review by ares — APPROVED** [review]
Ack pending tests not showing anything new, will merge right after.

**Review by ares — APPROVED** [review]
Thanks @xprazak2, merging!

**Pull request merged**